gobblin-metrics/src/main/java/gobblin/metrics/KafkaAvroReporter.java

```diff
@@ -0,0 +1,161 @@
+ */
+public class KafkaAvroReporter extends KafkaReporter {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(KafkaAvroReporter.class);
+
+  /**
+   * Avro schema used for encoding metrics
+   * TODO: finalize metrics avro schema
+   */
+  static String SchemaString = "{\n"+
+      " \"type\": \"record\",\n"+
+      " \"name\": \"Metric\",\n"+
+      " \"namespace\":\"<TBD>\",\n"+
+      " \"fields\" : [\n"+
+      " {\"name\": \"hostname\", \"type\": [\"null\", \"string\"], \"doc\": \"hostname emitting the metric\"},\n"+
+      " {\"name\": \"environment\", \"type\": [\"null\", \"string\"], \"doc\": \"environment where metric was emitted\"},\n"+
+      " {\"name\": \"tags\", \"type\": {\"type\": \"array\", \"items\": \"string\"}, \"doc\": \"tags associated with the metric\"},\n"+
+      " {\"name\": \"name\", \"type\": \"string\", \"doc\": \"metric name\"},\n"+
+      " {\"name\": \"value\", \"type\": [\"boolean\", \"int\", \"long\", \"float\", \"double\", \"bytes\", \"string\"], \"doc\": \"metric value\"}\n"+
+      " ]\n"+
+      "}";
+
+  Schema _schema;
+  private GenericDatumWriter<GenericRecord> w;
+  private Encoder e;
+  private ByteArrayOutputStream out;
+  private int exceptionCount;
+
+  protected KafkaAvroReporter(MetricRegistry registry, String name, MetricFilter filter, TimeUnit rateUnit,
+      TimeUnit durationUnit, String brokers, String topic, String host, String env, Set<String> tags) {
+    super(registry, name, filter, rateUnit, durationUnit, brokers, topic, host, env, tags);
+
+    Schema.Parser parser = new Schema.Parser();
+    _schema = parser.parse(SchemaString);
+
+    exceptionCount = 0;
+
+    out = new ByteArrayOutputStream();
+    e = EncoderFactory.get().binaryEncoder(out, null);
+    w = new GenericDatumWriter<GenericRecord>(_schema);
+
+  }
+
+  /**
+   * Serializes a metric key-value pair into an avro object
+   * Calls _serializeValue, which is synchronized because avro serializer is not thread-safe
+   *
+   * @param name name of the metric
+   * @param value value of the metric to report
+   * @param path additional suffixes to further identify the meaning of the reported value
+   * @return
+   */
+  @Override
+  protected byte[] serializeValue(String name, Object value, String... path) {
+    return _serializeValue(name, value, path);
+  }
+
+  protected synchronized byte[] _serializeValue(String name, Object value, String... path) {
+
+    GenericRecord record = new GenericData.Record(_schema);
+
+    record.put("name", makeName(name, path));
+    record.put("value", value);
+    record.put("hostname", _host);
+    record.put("environment", _env);
+    record.put("tags", _tags);
+
+    try {
+      out.reset();
+      w.write(record, e);
+      e.flush();
+      return out.toByteArray();
+    } catch(IOException e) {
+      exceptionCount++;
+      LOGGER.trace("Could not serialize Avro record for Kafka Metrics. Exception: %s", e.getMessage());
+      if (exceptionCount % 1000 == 0) {
```

Comment:
If there is a problem with the serialization, this exception would be thrown for every single metric. There could be hundreds of metrics, each being serialized every ten seconds (reasonable reporting interval).